devdoc/singlylinkedlist2_requirements.md

```diff
@@ -0,0 +1,176 @@
+singlylinkedlist2 requirements
+================
+
+## Overview
+
+SinglyLinkedList is module that provides the functionality of a singly linked list, allowing its user to add, remove and iterate the list elements.
+
+## Exposed API
+
+```c
+typedef struct SLIST_ENTRY_TAG
+{
+    struct SLIST_ENTRY_TAG *next;
+} SLIST_ENTRY, *PSLIST_ENTRY;
+
+typedef bool (*SLIST_MATCH_FUNCTION)(PSLIST_ENTRY listEntry, void* matchContext);
+typedef bool (*SLIST_CONDITION_FUNCTION)(PSLIST_ENTRY listEntry, void* matchContext, bool* continueProcessing);
+typedef int (*SLIST_ACTION_FUNCTION)(PSLIST_ENTRY listEntry, void* action_context, bool* continue_processing);
+
+MOCKABLE_FUNCTION(, void, SList_InitializeListHead, PSLIST_ENTRY, listHead);
+MOCKABLE_FUNCTION(, int, SList_IsListEmpty, const PSLIST_ENTRY, listHead);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_Add, PSLIST_ENTRY, listHead, PSLIST_ENTRY, listEntry);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_AddHead, PSLIST_ENTRY, listHead, PSLIST_ENTRY, listEntry);
+MOCKABLE_FUNCTION(, int, SList_Remove, PSLIST_ENTRY, listHead, PSLIST_ENTRY, listEntry);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_RemoveHead, PSLIST_ENTRY, listHead);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_GetNextItem, PSLIST_ENTRY, listEntry);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, SList_Find, PSLIST_ENTRY, listHead, SLIST_MATCH_FUNCTION, matchFunction, void*, matchContext);
+MOCKABLE_FUNCTION(, int, SList_RemoveIf, PSLIST_ENTRY, listHead, SLIST_CONDITION_FUNCTION, conditionFunction, void*, matchContext);
+MOCKABLE_FUNCTION(, int, SList_ForEach, PSLIST_ENTRY, listHead, SLIST_ACTION_FUNCTION, actionFunction, void*, actionContext);
+
+```
+## Usage
+
+`SLIST_ENTRY` does not store any data. Instead, `SLIST_ENTRY` is placed inside a struct:
+
+```c
+typedef struct DATA_TAG
+{
+    int num1;
+    int num2;
+    SLIST_ENTRY anchor;
+} DATA;
+```
+
+The parent struct of a list entry can be obtained using `CONTAINING_RECORD`:
+
+```
+PSLIST_ENTRY listEntry = SList_RemoveHead(listHead);
+DATA* data = CONTAINING_RECORD(listEntry, DATA, anchor);
+```
+
+### SList_InitializeListHead
+```c
+void SList_InitializeListHead(PSLIST_ENTRY listHead);
+```
+SList_InitializeListHead will initialize the listHead next pointer points to NULL. 
+
+### SList_IsListEmpty
+```c
+int SList_IsListEmpty(const PSLIST_ENTRY listHead);
```

Comment:
Is there any specific reason for the return type to be `int` instead of bool (true or false)?

Reply:
changed to bool